src/lib.rs

```diff
@@ -1,27 +1,50 @@
-use std::collections::HashMap;
-
 use borsh::schema::{BorshSchemaContainer, Declaration, Definition, Fields, VariantName};
 use schemars::schema::{RootSchema, Schema};
-use serde::{Deserialize, Serialize};
+use semver::Version;
+use serde::{de, Deserialize, Deserializer, Serialize};
+use std::collections::HashMap;
 
 #[doc(hidden)]
 #[cfg(feature = "__chunked-entries")]
 #[path = "private.rs"]
 pub mod __private;
 
 /// Current version of the ABI schema format.
-pub const SCHEMA_VERSION: &str = env!("CARGO_PKG_VERSION");
+pub const SCHEMA_VERSION_MAJOR: u64 = 0;
+pub const SCHEMA_VERSION_MINOR: u64 = 1;
+pub const SCHEMA_VERSION: &str = "0.1.0";
 
 /// Contract ABI.
 #[derive(Clone, Serialize, Deserialize, Debug, PartialEq)]
 pub struct AbiRoot {
     /// Semver of the ABI schema format.
-    #[serde(rename = "abi_schema_version")]
+    #[serde(deserialize_with = "ensure_current_version")]
     pub schema_version: String,
     /// Metadata information about the contract.
     pub metadata: AbiMetadata,
     /// Core ABI information (functions and types).
-    pub abi: AbiEntry,
+    pub body: AbiBody,
```

Comment:
Also, what if we just got rid of this entry in the ABI entirely and just flattened it into the root.

```json
{
  "schema_version": " ",
  "metadata": { },
  "functions": [ ],
  "root_schema": { }
}
```

Reply:
The reason why I decided against this in the first place to separate metadata from contents that have actual semantical impact. But this is discussable of course if you have an opinion on the matter.

Reply:
Not a strong opinion, just a thought. Seeing as it's a bit tricky finding a suitable name for the entry. Perhaps it's because it didn't need one. But it's fine nonetheless, we can retain it.